lib/lib_webgpu_dawn.cpp

```diff
@@ -1756,7 +1775,36 @@ void wgpu_buffer_map_async(WGpuBuffer buffer, WGpuBufferMapCallback callback, vo
 }
 
 void wgpu_buffer_map_sync(WGpuBuffer buffer, WGPU_MAP_MODE_FLAGS mode, double_int53_t offset, double_int53_t size) {
-  assert(false);
+  assert(wgpu_is_buffer(buffer));
+  
+  _WGpuObjectBuffer* obj = (_WGpuObjectBuffer*)_wgpu_get(buffer);
+  if (obj->state != kWebGPUBufferMapStateUnmapped)
+  	return;
+  
+  obj->state = kWebGPUBufferMapStatePending;
+  
+  struct _Data {
+  	WGpuBuffer buffer;
+  	WGPU_MAP_MODE_FLAGS mode;
+  	bool done = false;
+  };
+  
+  auto callback = [](WGPUBufferMapAsyncStatus status, void* rawData) {
+  	auto* userdata = static_cast<_Data*>(rawData);
+  	userdata->done = true;
+  	_Data* data = (_Data*)userdata;
+  	_WGpuObjectBuffer* obj = (_WGpuObjectBuffer*)_wgpu_get(data->buffer);
+  	obj->state = (data->mode & WGPUMapMode_Write) ? kWebGPUBufferMapStateMappedForWriting : kWebGPUBufferMapStateMappedForReading;
+  };
+  
+  _Data* data = new _Data{ buffer, mode };
```

Comment:
Does this need to be allocated with `new`? Or would a stack allocated `_Data data = { buffer, mode };` work? (good either way, just curious)

Reply:
I suppose this particular one could be a local var. It's a very low frequency function, so I don't know if putting it on the stack would save much.